### ISSUE TITLE
Using SuccessiveHalvingPruner

### ISSUE THREAD

**Issue by pvcastro** [himkt/allenopt#39]
Hi there!

I'm trying to use SuccessiveHalvingPruner, since some sources report it's a better alternative to ray's PBT. I was trying to leave the default attributes, but per my previous issue (#38 ), I had to specify something to test. First I tried the following:

```
{
    "pruner": {
        "type": "SuccessiveHalvingPruner",
        "attributes": {
            "min_resource": "auto"
        }
    }
}
```

But looks like it's getting parsed as None, so optuna's class is throwing an exception here:

![image](https://user-images.githubusercontent.com/12713359/117013165-fdb58a80-acc5-11eb-992b-1553ffa57551.png)

Then I tried using the following below, and it's working:

```
{
    "pruner": {
        "type": "SuccessiveHalvingPruner",
        "attributes": {
            "min_resource": 5
        }
    }
}
```

But what exactly is a "resource"? I don't understand the `min_resource` explanation from [optuna's documentation](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.pruners.SuccessiveHalvingPruner.html#optuna.pruners.SuccessiveHalvingPruner). 

Thanks!

**Comment by himkt** [comment]
Hello @pvcastro, please give me some time to investigate it.
(It should be a bug in `allennlp-optuna`, sorry for the inconvenience.)

**Comment by himkt** [comment]
Please upgrade allennlp-optuna and check if the error still happens.

> what exactly is a "resource"

`min_resource` controls a computation budget for each trial.
The smaller `min_resource` is, the more likely pruning will occur.
(ref. https://github.com/optuna/optuna/blob/d9a1f83a55953069ce322d0e8d778e91f0def525/optuna/pruners/_successive_halving.py#L182-L186)
If you want to learn more detail, please refer to the [paper](https://arxiv.org/abs/1810.05934) or ask Optuna dev team on [gitter](https://gitter.im/optuna/optuna).

**Comment by himkt** [comment]
@pvcastro Could you please upgrade allennlp-optuna and check if the error still happens?

> what exactly is a "resource"

`min_resource` controls a computation budget for each trial.
The smaller `min_resource` is, the more likely pruning will occur.

- ref. https://github.com/optuna/optuna/blob/d9a1f83a55953069ce322d0e8d778e91f0def525/optuna/pruners/_successive_halving.py#L182-L186

If you want to learn more detail, please refer to the [original paper](https://arxiv.org/abs/1810.05934) or ask Optuna dev team on [gitter](https://gitter.im/optuna/optuna).

**Comment by pvcastro** [comment]
Hi @himkt !
Leaving "auto" for min_resource is still throwing the same exception after upgrading to 0.1.6.

**Comment by himkt** [comment]
Finally, I found the root cause.

Optuna pruning integration for AllenNLP creates a pruner using environment variables.
If a value `None` (None object) is specified in config, it would be parsed as "None" (string literal) when constructing a pruner. It causes the problem you encountered.

```
(Pdb) _get_environment_variables_for_pruner()
{'min_resource': 'None', 'reduction_factor': 4, 'min_early_stopping_rate': 0}
```

https://github.com/optuna/optuna/blob/master/optuna/integration/allennlp.py#L74-L95

I'll make a PR to avoid this problem.
But for now, could you please specify `min_resource` to be `"auto"` explicitly in your configuration?

**Comment by himkt** [comment]
@pvcastro

**Comment by pvcastro** [comment]
Hi @himkt , as I stated [here](https://github.com/himkt/allennlp-optuna/issues/39#issuecomment-840869459), leaving "auto" still causes the issue. The "auto" value is being ignored, for some reason.

**Comment by pvcastro** [comment]
I was doing some debugging, and I think the problem is that at [this line at optuna](https://github.com/optuna/optuna/blob/853b94b459ecf9e434fb1e28dda7f07349eec20c/optuna/pruners/_successive_halving.py#L156), they don't save the "auto" value, they only save a value if it's an integer. And whey you retrieve the pruner attributes from the environment variables in the method you mentioned, the min_resource is None, and when you try to instantiate a new one with None, it throws the exception.

**Comment by himkt** [comment]
Oh, you're right, thank you.

I now understand the problem is happened due to the mismatch between `min_resource` and `_min_resource`.
As you pointed out, SuccessiveHalvingPruner doesn't preserve the `min_resource` when it is not integer.
I'll discuss how to restore a pruner from attributes such as `_min_resource` (slightly different from the original `min_resource`) with Optuna dev team, sorry but please wait. :bow: (thank you so much for kindly and precise investigations)

**Comment by pvcastro** [comment]
Sure, glad to help!

**Comment by himkt** [comment]
Sorry for being late @pvcastro. I resolved this problem in https://github.com/optuna/optuna/pull/2731.
Please run `pip uninstall optuna` and `run pip install 'optuna@git+https://github.com/optuna/optuna@master'` on your environment. (This fix will be included in the next version of Optuna)

**Comment by himkt** [comment]
Closed as this fix was shipped in Optuna v2.9.0.
